pkg/compactor/blocks_cleaner.go

```diff
@@ -410,15 +424,39 @@ func (c *BlocksCleaner) cleanUser(ctx context.Context, userID string, userLogger
 		}
 	}()
 
-	// Read the bucket index.
+	retries := backoff.New(ctx, backoff.Config{
+		MaxRetries: 5,
+	})
+	var lastErr error
+
+	for retries.Ongoing() {
```

Comment:
Would it be better to retry the individual operation that failed, rather than the entire cleanUser flow?

Reply:
> pursue the "RetryingBucket" idea
I'll workshop that in a separate PR and close this one if it works out.

Reply:
Something like a `RetryingBucket` implies retries are done internally to the bucket, but even if we have to externally retry bucket operations, my comment was just about whether this is better - to retry just the operation that failed - rather than all of the operations in cleanUser. Your point about idempotency makes sense though.

Reply:
I prototyped both.

* RetryingBucket...Client https://github.com/grafana/mimir/pull/8052
* higher-level (but not that high) operation retry  https://github.com/grafana/mimir/pull/8071

The retryingBucket is kinda neat but ultimately it is not a great abstraction because you have to think about how much time you're going to give to specific operations... but store that info at the client layer and pass that into some function that is going to do 4 different operations across 30 different blob store objects. Kind of an impedance mismatch. I think the second one there is the one we should go with.